2019/day3/plane.go

```diff
@@ -0,0 +1,139 @@
+	return p.objectID
+}
+
+func (p *planePoint) setObjectID(id int) {
+	p.objectID = id
+}
+
+func (p *planePoint) getOrder() int {
+	return p.order
+}
+
+// Plane represents an x,y coordinate plane.
+// A plane can plot many different objects and it assigns each new object
+// an objectID to tell them apart.
+//
+// The x,y coordinate plane is represented by a map defined as:
+//   - key: an x,y point on the plane
+//   - value: list of object's orderedPoints at that position on the plane
+//
+// The centralPoint is the start location of all objects when they
+// are plotted on the plane.
+type Plane struct {
+	centralPoint               Point
+	closestIntersection        int
+	closestIntersectionByOrder int
+	points                     map[Point][]planePoint
+	objectIDCounter            int
+}
+
+func NewPlane(centralPoint Point) *Plane {
+	return &Plane{
+		centralPoint:               centralPoint,
+		closestIntersection:        math.MaxInt32,
+		closestIntersectionByOrder: math.MaxInt32,
+		points:                     map[Point][]planePoint{},
+	}
+}
+
+// Plot plots a new object on the plane.
+// An object is represented by a collection of planePoints.
+// When a new object is plotted all the planePoints get the same objectID.
+func (p *Plane) Plot(points []planePoint) {
+	p.objectIDCounter++
+	for _, point := range points {
+		point.setObjectID(p.objectIDCounter)
+		p.add(point)
+	}
+}
+
+// add adds a new x,y point to the plane. When a new point is added to
+// the plane, it checks if it intersects with any other objects. If the new
+// point does intersect with other objects, then it calculates if its the
+// closest intersection to the central point.
+func (p *Plane) add(newPoint planePoint) {
+	existingPoints, ok := p.points[newPoint.getPoint()]
+	if ok {
+		for _, existingPoint := range existingPoints {
+			if p.pointsIntersect(existingPoint, newPoint) {
+				// if the newPoint intersects with an existingPoint then check if the distance
+				// of this intersection is closer than any other intersection to the central point.
+				distance := manhattenDistance(p.centralPoint, newPoint.getPoint())
+				if distance < p.closestIntersection {
+					p.closestIntersection = distance
+				}
+
+				distanceByOrder := existingPoint.order + newPoint.order
+				if distanceByOrder < p.closestIntersectionByOrder {
+					p.closestIntersectionByOrder = distanceByOrder
+				}
+			}
+		}
+	}
+	p.points[newPoint.getPoint()] = append(p.points[newPoint.getPoint()], newPoint)
+}
+
+func (p *Plane) pointsIntersect(point1, point2 planePoint) bool {
+	if point1.getObjectID() == point2.getObjectID() {
+		return false
+	}
+	if p.isCentralPoint(point1) || p.isCentralPoint(point2) {
+		return false
+	}
+	return point1.x() == point2.x() && point1.y() == point2.y()
+}
+
+func (p *Plane) isCentralPoint(point planePoint) bool {
+	return point.x() == p.centralPoint.x() && point.y() == p.centralPoint.y()
+}
+
+func manhattenDistance(point1, point2 Point) int {
```

Comment:
I usually like to have methods like `Add`, `Sub` on such `Point`/`Vector` types. For example see https://pkg.go.dev/image#Point. They end up condensing the code quite a bit. For example, in this case it could be:

```
p := point1.Sub(point2).Abs()
return p.X + p.Y
```

Reply:
Also a typo `manhattanDistance` :)